deepinv/optim/distance.py

```diff
@@ -0,0 +1,364 @@
+              \right.
+
+
+    :param float radius: radius of the ball. Default: None.
+
+    """
+
+    def __init__(self, radius=None):
+        super().__init__()
+        self.radius = radius
+
+    def fn(self, x, y, *args, radius=None, **kwargs):
+        r"""
+        Computes the batched indicator of :math:`\ell_2` ball with radius `radius`, i.e. :math:`\iota_{\mathcal{B}(y,r)}(x)`.
+
+        :param torch.Tensor x: Variable :math:`x` at which the indicator is computed. :math:`u` is assumed to be of shape (B, ...) where B is the batch size.
+        :param torch.Tensor y: Observation :math:`y` of the same dimension as :math:`u`.
+        :param float radius: radius of the :math:`\ell_2` ball. If `radius` is None, the radius of the ball is set to `self.radius`. Default: None.
+        :return: (torch.Tensor) indicator of :math:`\ell_2` ball with radius `radius`. If the point is inside the ball, the output is 0, else it is 1e16.
+        """
+        diff = x - y
+        dist = torch.norm(diff.reshape(diff.shape[0], -1), p=2, dim=-1)
+        radius = self.radius if radius is None else radius
+        loss = (dist > radius) * 1e16
+        return loss
+
+    def prox(self, x, y, *args, radius=None, gamma=None, **kwargs):
+        r"""
+        Proximal operator of the indicator of :math:`\ell_2` ball with radius `radius`, i.e.
+
+        .. math::
+
+            \operatorname{prox}_{\iota_{\mathcal{B}_2(y,r)}}(x) = \operatorname{proj}_{\mathcal{B}_2(y, r)}(x)
+
+
+        where :math:`\operatorname{proj}_{C}(x)` denotes the projection on the closed convex set :math:`C`.
+
+
+        :param torch.Tensor x: Variable :math:`x` at which the proximity operator is computed.
+        :param torch.Tensor y: Observation :math:`y` of the same dimension as :math:`x`.
+        :param float gamma: step-size. Note that this parameter is not used in this function.
+        :param float radius: radius of the :math:`\ell_2` ball.
+        :return: (torch.Tensor) projection on the :math:`\ell_2` ball of radius `radius` and centered in `y`.
+        """
+        radius = self.radius if radius is None else radius
+        diff = x - y
+        dist = torch.norm(diff.reshape(diff.shape[0], -1), p=2, dim=-1)
+        return y + diff * (
+            torch.min(torch.tensor([radius]).to(x.device), dist) / (dist + 1e-12)
+        ).view(-1, 1, 1, 1)
+
+
+class KullbackLeiblerDistance(Distance):
+    r"""
+    "Generalized" Kullback-Leibler divergence. It is also the log-likelihood of the Poisson distribution.
+
+    .. math::
+
+        \d{y}{x} =  \sum_i y_i \log(y_i / x_i) + x_i - y_i
+
+
+    .. note::
+
+        The function is not Lipschitz smooth w.r.t. :math:`x` in the absence of background (:math:`\beta=0`).
+
+    :param float gain: gain of the measurement :math:`y`. Default: 1.0.
+    :param float bkg: background level :math:`\beta`. Default: 0.
+    :param bool normalize: if True, the measurement is multiplied by the gain. Default: True.
```

Comment:
same here regarding 'denormalize'. Maybe we can also put a warning linking to the PoissonNoise class, saying that by default the poisson noise normalizes the measurements.